python_modules/dagster/dagster/core/execution/plan/compute_generator.py

```diff
@@ -58,151 +70,151 @@ def compute(context, input_defs) -> Generator[Output, None, None]:
 
 async def _coerce_async_solid_to_async_gen(awaitable, context, output_defs):
     result = await awaitable
-    for event in _validate_and_coerce_solid_result_to_iterator(result, context, output_defs):
+    for event in validate_and_coerce_solid_result_to_iterator(result, context, output_defs):
         yield event
 
 
 def _coerce_solid_compute_fn_to_iterator(fn, output_defs, context, context_arg_provided, kwargs):
     result = fn(context, **kwargs) if context_arg_provided else fn(**kwargs)
-    for event in _validate_and_coerce_solid_result_to_iterator(result, context, output_defs):
+    for event in validate_and_coerce_solid_result_to_iterator(result, context, output_defs):
         yield event
 
 
-def _validate_and_coerce_solid_result_to_iterator(result, context, output_defs):
+def _zip_and_iterate_solid_result(
+    result: Any, context: OpExecutionContext, output_defs: Sequence[OutputDefinition]
+) -> Iterator[Tuple[int, Any, OutputDefinition]]:
+    if len(output_defs) > 1:
+        if (
+            context.solid_def.is_from_decorator()
+            and context.solid_def.compute_fn.get_output_annotation()
+            and get_origin(context.solid_def.get_output_annotation()) == tuple
+        ):
+            for position, (output_def, element) in enumerate(zip(output_defs, result)):
+                yield position, output_def, element
+        else:
+            raise DagsterInvariantViolationError(
+                f"{context.describe_op()} has multiple outputs, but only one "
+                f"output was returned of type {type(result)}. When using "
+                "multiple outputs, either yield each output, or return a tuple "
+                "containing a value for each output. Check out the "
+                "documentation on outputs for more: "
+                "https://docs.dagster.io/concepts/ops-jobs-graphs/ops#outputs."
+            )
+    else:
+        yield 0, output_defs[0], result
```

Comment:
zip up the outputs with their accompanying output defs so that output checking logic can be reused across the single and multi output case (before they were split).